app/uk/gov/hmrc/gform/sharedmodel/DataRetrieve.scala

```diff
@@ -0,0 +1,131 @@
+
+object DataRetrieveId {
+  implicit val format: Format[DataRetrieveId] =
+    JsonUtils.valueClassFormat[DataRetrieveId, String](DataRetrieveId.apply, _.value)
+
+  val idValidation: String = "[_a-zA-Z]\\w*"
+  val unanchoredIdValidation: Regex = s"""$idValidation""".r
+}
+
+sealed trait DataRetrieveAttribute {
+  def exprId: String
+}
+case object DataRetrieveIsValid extends DataRetrieveAttribute {
+  override def exprId: String = "isValid"
+}
+
+case object DataRetrieveAttribute {
+
+  implicit val format: OFormat[DataRetrieveAttribute] = derived.oformat()
+
+  val idValidation: String = "[_a-zA-Z]\\w*"
+  val unanchoredIdValidation: Regex = s"""$idValidation""".r
+
+  def fromString(value: String): DataRetrieveAttribute = value match {
+    case "DataRetrieveIsValid" => DataRetrieveIsValid
+    case other                 => throw new IllegalArgumentException(s"Unknown DataRetrieveAttribute value $other")
+  }
+
+  def fromExpr(exprId: String): DataRetrieveAttribute = exprId match {
+    case "isValid" => DataRetrieveIsValid
+    case other     => throw new IllegalArgumentException(s"Unknown DataRetrieveAttribute value $other")
+  }
+}
+
+sealed trait DataRetrieve {
+  def id: DataRetrieveId
+  def attributes: List[DataRetrieveAttribute]
+  def formCtxExprs: List[Expr]
+}
+case class ValidateBank(override val id: DataRetrieveId, sortCode: Expr, accountNumber: Expr) extends DataRetrieve {
+  override def attributes: List[DataRetrieveAttribute] = List(DataRetrieveIsValid)
+  override def formCtxExprs: List[Expr] = List(sortCode, accountNumber)
+}
+
+object DataRetrieve {
+
+  val reads: Reads[DataRetrieve] = new Reads[DataRetrieve] {
+    override def reads(json: JsValue): JsResult[DataRetrieve] =
+      (for {
+        typeValue <- opt[String](json, "type")
+        idValue   <- opt[String](json, "id")
+        dataRetrieve <- typeValue match {
+                          case "validateBankDetails" =>
+                            for {
+                              parameters         <- opt[JsObject](json, "parameters")
+                              sortCodeValue      <- opt[String](parameters, "sortCode")
+                              accountNumberValue <- opt[String](parameters, "accountNumber")
+                              sortCodeExpr       <- BasicParsers.validateWithParser(sortCodeValue, ValueParser.expr)
+                              accountNumberExpr  <- BasicParsers.validateWithParser(accountNumberValue, ValueParser.expr)
+                            } yield ValidateBank(DataRetrieveId(idValue), sortCodeExpr, accountNumberExpr)
+                          case other => Left(UnexpectedState(s"'type' value $other not recognized"))
+                        }
+      } yield dataRetrieve).fold(e => JsError(e.error), r => JsSuccess(r))
+  }
+
+  def opt[T](jsValue: JsValue, path: String)(implicit r: Reads[T]): Opt[T] =
+    jsValue \ path match {
+      case JsDefined(json) =>
+        Try(json.as[T])
```

Comment:
There is `validate` method for checking if json can be converted to  type `T`

Reply:
Replaced with validate method